### ISSUE TITLE
Prevent/handle concurrent deployments

### ISSUE THREAD

**Issue by Al2Klimov** [Icinga/icingaweb2-module-director#2412]
## Expected Behavior

At best a deployment waits while others are running.

## Current Behavior

* https://community.icinga.com/t/handling-of-parallel-deploy-actions/8433
* https://community.icinga.com/t/multiple-deploys-at-the-same-time-will-cause-comments-scheduled-downtime-loss/8428

## Your Environment

**Comment by Thomas-Gelf** [comment]
The Deployment request takes less then a second up to a very few seconds, but the Deployment carried out by Icinga can take minutes. There is no safe way of knowing whether a deployment is still ongoing, failed - or will never return at all. That's why Config Packages are deployed to isolated directories, with dedicated outcome and startup log.

* if we mess up Scheduled Downtimes and Comments in such scenario (as suggested in one of the above forum posts), I'd consider this being a **severe** bug
* when Icinga get's a second deployment request, while still being busy with the former one, it could either reject or accept the request. Currently it would accept the request and probably interrupt the ongoing former deployment - which is IMHO the correct behaviour
* if an application hammers the system with lot's of harmful parallel deployment requests, Director behaves as how it's being told to do
* using the Deployment Job in the Director helps greatly: you can auto-deploy with a very short interval, and add a reasonable grace period - just to be on the safe side. This is IMHO currently the best solution. In large environments with hundreds of changes a day, it makes it easy to bring deployments down to just a very few per day
* alternatively, the deployment logic could be optimized by using the information provided by the [Deployment Status API Request](https://github.com/Icinga/icingaweb2-module-director/blob/master/doc/70-REST-API.md#deployment-status). You can use the checksum returned by `director/config/deploy` and compare it to `director/config/deployment-status`. This would of course require a coordinated way of steering those API request. No chance of doing so, when you have hundreds of systems firing lots of autonomous deploy requests. But that's an imaginary nightmare scenario, isn't it? :rofl: 

Long story short: Director is only the "transport" for these requests, and shouldn't take the responsibility for deciding whether Icinga is in the right mood to accept a deployment or not ;-)

**Comment by Thomas-Gelf** [comment]
> Huh? If you fire a deploy, isn't there a spinner turning sooner or later into a tick/x along with an Icinga validation log?

Yes, and as long as it is spinning, the background daemon tries to collect related information again and again. However, sometimes Icinga completely "forgets" about a Deployment. Probably when deploying twice, when being restarted by other means (Puppet, Ansible) or whatever. Then it would wait forever, and older versions of Director did so.

Recent versions apply the following logic: in case a more recent configuration has been deployed, and Icinga shows an outcome (status, startup.log) for that deployment, all former ones (which still show no outcome) are considered being lost. Then Director marks them as such in it's DB and wipes the orphaned stages via the Icinga API.

So, when Director would stop deploying as requested, the first lost Deployment would result in a deadlock situation.

**Comment by julianbrost** [comment]
Has anyone actually looked at this in more detail? Like where exactly are things missing? Only in IDO or also in icinga2 itself (i.e. are they missing in icinga2 API responses)? So sounds like this could also be an IDO bug when restarting in the wrong moment or something like that.

**Comment by Thomas-Gelf** [comment]
Unless Icinga refuses deployments (preferably with a dedicated error message and/or status code) while being in a busy state, there is nothing we can do here I guess. Closing the feature request for now, please do not hesitate to create a new one (or re-open this one) in case this changes.
